api/v1alpha1/healthcheck_types.go

```diff
@@ -0,0 +1,163 @@
+	// HealthChecker defines the concrete health checker to do health checking.
+	HealthChecker HealthChecker `json:"healthChecker"`
+}
+
+// HealthCheckerType is the type of health checker.
+// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+type HealthCheckerType string
+
+const (
+	// HealthCheckerTypeHTTP defines the HTTP type of health checking.
+	HealthCheckerTypeHTTP HealthCheckerType = "HTTP"
+	// HealthCheckerTypeGRPC defines the GRPC type of health checking.
+	HealthCheckerTypeGRPC HealthCheckerType = "GRPC"
+	// HealthCheckerTypeTCP defines the TCP type of health checking.
+	HealthCheckerTypeTCP HealthCheckerType = "TCP"
+)
+
+// HealthChecker defines the configuration of concrete health checker.
+// EG supports various types of health checking including HTTP, GRPC, TCP.
+// +union
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'HTTP' ? has(self.http) : !has(self.http)",message="If Health Checker type is HTTP, http field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'GRPC' ? has(self.grpc) : !has(self.grpc)",message="If Health Checker type is GRPC, grpc field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'TCP' ? has(self.tcp) : !has(self.tcp)",message="If Health Checker type is TCP, tcp field needs to be set."
+type HealthChecker struct {
+	// Type defines the type of health checker.
+	// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+	// +unionDiscriminator
+	Type HealthCheckerType `json:"type" yaml:"type"`
+	// HTTP defines the configuration of http health checker.
+	// It's required while the health checker type is HTTP.
+	// +optional
+	HTTP *HTTPHealthChecker `json:"http,omitempty" yaml:"http,omitempty"`
+	// GRPC defines the configuration of grpc health checker.
+	// It's required while the health checker type is GRPC.
+	// +optional
+	GRPC *GRPCHealthChecker `json:"grpc,omitempty" yaml:"grpc,omitempty"`
+	// TCP defines the configuration of tcp health checker.
+	// It's required while the health checker type is TCP.
+	// +optional
+	TCP *TCPHealthChecker `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
+	// Method defines the HTTP method used for health checking.
+	// Defaults to GET
+	// +optional
+	Method *string `json:"method,omitempty" yaml:"method,omitempty"`
+	// ExpectedStatuses defines a list of HTTP response statuses considered healthy.
+	// +optional
+	ExpectedStatuses []HTTPStatusRange `json:"expectedStatuses,omitempty" yaml:"expectedStatuses,omitempty"`
+	// ExpectedResponses defines a list of HTTP expected responses to match.
+	// +optional
+	ExpectedResponses []HealthCheckPayload `json:"expectedResponses,omitempty" yaml:"expectedResponses,omitempty"`
```

Comment:
why is this a list ? is this a OR match ?

Reply:
According to [Envoy doc](https://www.envoyproxy.io/docs/envoy/v1.28.0/api-v3/config/core/v3/health_check.proto#envoy-v3-api-msg-config-core-v3-healthcheck-tcphealthcheck), it's a AND match. 

Perhaps in most cases, we only need one?

Reply:
can you think of use cases where we'd need multiple matches ?

Reply:
not really.

Reply:
lets make it singular then

Reply:
done